BALSAMIC/utils/cli.py

```diff
@@ -397,22 +400,133 @@ def get_bioinfo_tools_version(
     return bioinfo_tools_version
 
 
+def validate_fastq_input(sample_dict: dict, fastq_path: str) -> None:
+    """Validates fastq input in sample dict and fastq_path.
+
+    Criteria:
+        - All fastq-files in the supplied fastq_dir must have been assigned to a sample.
+        - All fastq-files in the sample dict must exist.
+
+    Args:
+        sample_dict: dict. Dictionary of sample-names and their associated sample-type and fastq-information.
+        fastq_path: str. Path to the directory containing the fastq-files.
+
+    Returns:
+        Nothing
+    """
+    complete_fastq_list = glob.glob(f"{fastq_path}/*fastq.gz")
+    assigned_fastq_list = []
+    for sample in sample_dict:
+        fastq_dict = sample_dict[sample]["fastq_info"]
+        for fastq_pattern in fastq_dict:
+            for fastq_read_direction, fastq_path in fastq_dict[fastq_pattern].items():
+                assigned_fastq_list.append(fastq_path)
+                if not Path(fastq_path).is_file():
+                    error_message = f"Fastq-file does not exist: {fastq_path}"
+                    LOG.error(error_message)
+                    raise FileNotFoundError(error_message)
+
+    unassigned_fastqs = []
+    for fastq in complete_fastq_list:
+        if fastq not in assigned_fastq_list:
+            unassigned_fastqs.append(fastq)
+
+    if unassigned_fastqs:
+        unassigned_fastqs_str = "\n".join(unassigned_fastqs)
+        error_message = f"Fastq files found in fastq directory not assigned to any sample: {unassigned_fastqs_str}"
+        LOG.error(error_message)
+        raise BalsamicError(error_message)
+
+
+def get_fastq_info(sample_name: str, fastq_path: str) -> Dict[str, str]:
+    """Returns a dictionary of fastq-patterns and fastq-paths existing in fastq_dir for a given sample.
+
+    Args:
+        sample_name: str. The name of the sample for which fastq-files will be searched for in the fastq_path.
+        fastq_path: str. Path to where the fastq-files should be found for the supplied sample_name.
+
+    Returns:
+        fastq_dict: dict. Nested dictionary containing the unique file-name pattern for each fastq-pair,
+        and the paths to the fastq-files.
+    """
+
+    fastqs_in_fastq_path = glob.glob(f"{fastq_path}/*fastq.gz")
+    if not fastqs_in_fastq_path:
+        error_message = f"No fastq files found in supplied fastq-path: {fastq_path}"
+        LOG.error(error_message)
+        raise FileNotFoundError(error_message)
+
+    fastq_dict = {}
+    fastq_suffixes = FASTQ_SUFFIXES
+    for suffix in fastq_suffixes:
+        fwd_suffix = fastq_suffixes[suffix]["fwd"]
+        rev_suffix = fastq_suffixes[suffix]["rev"]
+
+        fastq_fwd_regex = re.compile(
+            r"(^|.*_)" + sample_name + r"_.*" + fwd_suffix + r"$"
+        )
+        fwd_fastqs = [
+            f"{fastq_path}/{fastq}"
+            for fastq in os.listdir(fastq_path)
+            if fastq_fwd_regex.match(fastq)
+        ]
```

Comment:
Couldn't we just use glob instead, so it's easier to follow?

```suggestion
fwd_fastqs = glob.glob(os.path.join(fastq_path, f"*_{sample_name}_*{fwd_suffix}"))

```

Reply:
I think I switched to re because I wanted to allow for the possibility that the samplename in the fastq-name string could be either at the start of the name or separated by underscores, and I don't know if that is possible with glob. 

I could use:
fwd_fastqs = glob.glob(os.path.join(fastq_path, f"*_{sample_name}_*{fwd_suffix}"))
Which would require that samplename is separated by underscores
Or:
fwd_fastqs = glob.glob(os.path.join(fastq_path, f"*{sample_name}_*{fwd_suffix}"))
Which would allow for start of file, but also anything else. 

For now I'll keep the re solution unless you have a better suggestion : )

Reply:
I would say if we are already defining the fastqs name structure, then we should stick with it and only allow that format.  We are getting the sample name between underscores from CG, right? Then I think that's our format

Reply:
That's true, and there are some bonuses to having a predefined fastq-name structure. Like for instance we could extract lane and flow-cell easily from the fastq-name: "1_171015_HKL7FDMXX_ACC123456_XXXXXX_R_1", taking 1 and HKL7FDMXX and use it as meta-data. But maybe we can have a discussion about this and add this metadata in a separate PR along with this refactoring, which comes with making the pipeline a bit more annoying to use for external people (but this is ...probably not relevant).

Reply:
I think we should simplify this regex in any case. I would make it general for now as you say and adjust it to something more specific as you say in the future